rosdoc2/verbs/build/builders/doxygen_builder.py

```diff
@@ -243,7 +244,8 @@ def build(self, *, doc_build_folder, output_staging_directory):
         logger.info(
             f"Running Doxygen: '{' '.join(cmd)}' in '{working_directory}'"
         )
-        completed_process = subprocess.run(cmd, cwd=working_directory)
+        completed_process = subprocess.run(
+            cmd, cwd=working_directory, stdout=sys.stdout, stderr=sys.stderr)
```

Comment:
Isn't this doing the default behavior with more complexity. The default is that subprocess.run doesn't capture the output which means it goes to the console. 

```
>>> import sys
>>> subprocess.run('echo hi', shell=True, stdout=sys.stdout, stderr=sys.stderr)
hi
CompletedProcess(args='echo hi', returncode=0)
>>> subprocess.run('echo hi', shell=True)
hi
CompletedProcess(args='echo hi', returncode=0)
```

Reply:
The reason that I did this is so that I can remap sys.stdout and sys.stderr to a file, and capture the output of individual package runs. IIRC from my tests, without this I could not remap to capture the output.

I was definitely following my "Programming is an Experimental Science"™ philosophy here though, trying things until something worked. But I'll look this over again, and confirm that I really need to do this. It seems strange to me as well that this is needed, but it seems to work.

Reply:
If you do find that this helps with that extra bit of debuggability, it's fine to keep it. It would be great if you could add a comment to explain it so others can also find this benefit too.

Reply:
I tried this again, and I can confirm that leaving ```completed_process = subprocess.run(cmd, cwd=working_directory)``` without adding ```, stdout=sys.stdout, stderr=sys.stderr``` does not work. The outputs go to the console, which is not what we want, we want to capture to a file.

I tried looking at pytest to see how they do it, but got lost in the levels of abstraction. Pytest has a lot of code around capturing, but I am sure they deal with a lot of edge cases that we don't care about.

If I look at subprocess, the code says:
```
def run(*popenargs,
        input=None, capture_output=False, timeout=None, check=False, **kwargs):
    """Run command with arguments and return a CompletedProcess instance.

    The returned instance will have attributes args, returncode, stdout and
    stderr. By default, stdout and stderr are not captured, and those attributes
    will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them,
    or pass capture_output=True to capture both.
```
So in fact the default for syserr is not ```sys.stderr``` but None. We want to capture the output to our own PIPE, which we store in ```sys.stderr```.

We would not have to store that pipe in ```sys.stderr``` in the subprocess, we could define a variable for it,  but by doing so we also capture output generated by, for example, ```print(f"[rosdoc2] adding markdown parser", file=sys.stderr)```

Reply:
Sorry I missed that you're changing the sys.stdout/err in the per package scan code here https://github.com/ros-infrastructure/rosdoc2/pull/139/files#diff-4d67c84fa514f2538371532c7f2d3979b42da8852649e61ed9cac143e83f1c55R169-R180

Which is capturing the output different than capturing it through the subprocess native mechanisms. Instead of changing sys.stderr at the global level, I think it would make sense to extend the main_impl to include the ability to pass an alternative output file. And if that output file is set, then pass that through to the subprocess invocation. This will avoid worries about race conditions on globally redirecting for potential other implementations of parallel threading.

Reply:
I'm having a hard time understanding and implementing what you want here.

First, I don't see any mechanism within Pool to control capture of stdout and stderr, unlike the basic subprocess.run(). Do you know how to do that?

Once I get into package_impl, output is going to the normal sys.stdout and sys.stderr, and we have to fix that (there is no subprocess.run involved with build_main_impl). I fixed if via ```sys.stdout = outfile```. I don't see what else to do.

I was uncertain if that was the correct approach, but then I ran across contextlib.redirect_stdout in the standard library, and they do exactly what process_impl does: save the old value of sys.stdout, change sys.stdout to the new value, then restore it when they are done.

So I need more clarity about what you want here.